lib/repack.c

```diff
@@ -135,6 +137,39 @@ must_be_superuser(const char *func)
 #define LIBRARY_VERSION "unknown"
 #endif
 
+/*
+ * Try acquire an ACCESS EXCLUSIVE table lock, avoiding deadlocks and long
+ * waits.
+ * Arguments:
+ *
+ *  nspname: name of namespace of relation
+ *  relanme: name of relation
+ */
+static void
+lock_exclusive_wait(const char *nspname, const char *relname)
+{
+	int			i;
+
+	for (i = 1; ; i++)
+	{
+		int			wait_msec;
+
+		CHECK_FOR_INTERRUPTS();
+
+		/* wait for a while to lock the table. */
+		wait_msec = Min(1000, i * 100);
+
+		execute_with_format(SPI_OK_UTILITY,
+			"SET statement_timeout = %d", wait_msec);
```

Comment:
I don't think that would make a huge difference, but we could:
```suggestion
		if (i <= 1000)
		{
			wait_msec = i * 100;

			execute_with_format(SPI_OK_UTILITY,
				"SET statement_timeout = %d", wait_msec);
		}
```

To avoid unnecessary variable assignment and `SET` calls once we reach `1000`.

Reply:
Indeed it is an improvement.

But actually the function `lock_exclusive_wait` is broken. It runs infinitely :)

Secondly we cannot fix the issue within `repack_drop` function. The problem is that `repack_drop` is run in one transaction, and if any statement fails then the whole transaction needs to be rollbacked. If `LOCK TABLE` statement fails by `statement_timeout` then Postgres raises an error and the transaction needs to be rollbacked too.

It is a caller of `repack_drop` responsibility to set `statement_timeout` (or `lock_timeout`) and try again if it isn't succeed. `repack_drop` is called in `pg_repack.c`.

Reply:
But actually we can use SAVEPOINT. In that case we can fix it within `repack_drop`.